### PR TITLE
Thrift LZO Support for HIVE

### PR DESCRIPTION
**Pull request by mghosh4**
Read support for Thrift LZO Files

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        List<ObjectInspector> fieldObjectInspectors = new ArrayList<ObjectInspector>(
+            fieldTypeInfos.size());
+        for (int i = 0; i < fieldTypeInfos.size(); i++) {
+          fieldObjectInspectors
+              .add(getObjectInspector(fieldTypeInfos.get(i), resolver.getNestedResolver(i)));
+        }
+        result = new ThriftGeneralObjectInspector(fieldNames, fieldObjectInspectors, resolver);
+        break;
+      }
+      case UNION: {
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java:135]: Although, we translate all UNION to STRUCT when building meta info and this part of the code should not be hit. The UNION should actually be treated similarly as STRUCT. 

https://github.com/apache/thrift/blob/master/doc/specs/thrift-binary-protocol.md
https://github.com/apache/thrift/blob/master/doc/specs/thrift-compact-protocol.md

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+public class ThriftGeneralObjectInspector extends BaseStructObjectInspector {
+
+  ThriftFieldIdResolver resolver;
+
+  public ThriftGeneralObjectInspector(List<String> fieldNames, List<ObjectInspector> loi, ThriftFieldIdResolver resolver) {
+    super(fieldNames, loi);
+    this.resolver = resolver;
+  }
+
+  public Object getStructFieldData(Object data, StructField fieldRef)
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java]: Use `@Override` annotation

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  {
+    ThriftGenericRow row = (ThriftGenericRow) data;
+    return row.getFieldValueForThriftId(resolver.getThriftId(fieldRef.getFieldID()));
+  }
+
+  public List<Object> getStructFieldsDataAsList(Object data)
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java]: `@Override`

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralInputFormat.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    implements CombineHiveInputFormat.AvoidSplitCombination
+{
+  public ThriftGeneralInputFormat()
+  {
+    super(new MultiInputFormat());
+  }
+
+  private void initialize(FileSplit split, JobConf job) throws IOException
+  {
+    String thriftClassName = job.get(serdeConstants.SERIALIZATION_CLASS);
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralInputFormat.java]: Do we need to explicitly put a condition check to test `thriftClassName` is `null`?

File: ql/src/java/org/apache/hadoop/hive/ql/plan/PartitionDesc.java [hunk 5]
```diff
@@ -174,6 +174,11 @@ public String getDeserializerClassName() {
   public Deserializer getDeserializer(Configuration conf) throws Exception {
     Properties schema = getProperties();
     String clazzName = getDeserializerClassName();
+    if (tableDesc.getProperties() != null &&
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/plan/PartitionDesc.java]: I'm imaging you can use `schema` to replace `tableDesc.getProperties()` in this condition express.

```
if (schema != null && 
    "com.facebook.presto.twitter.hive.thrift.ThriftGenericRow".equals(schema.getProperty(serdeConstants.SERIALIZATION_CLASS)) {
```

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Table.java [hunk 6]
```diff
@@ -308,7 +309,13 @@ public HiveStorageHandler getStorageHandler() {
             return null;
           }
           inputFormatClass = getStorageHandler().getInputFormatClass();
+          if ("com.twitter.elephantbird.mapred.input.HiveMultiInputFormat".equals(inputFormatClass.getName())) {
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Table.java]: I think we won't come into this branch. Will we?

File: metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreUtils.java [hunk 7]
```diff
@@ -433,6 +433,9 @@ public static Deserializer getDeserializer(Configuration conf,
       org.apache.hadoop.hive.metastore.api.Table table, boolean skipConfError,
       String lib) throws MetaException {
     try {
+      if (table.getParameters().containsKey(serdeConstants.SERIALIZATION_CLASS) &&
```
Comment (Yaliang) [inline metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreUtils.java:436]: This code is only referred by DDLTask. I believe we don't need to change this code?

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  @Override
+  public ObjectInspector getObjectInspector() throws SerDeException {
+    return cachedObjectInspector;
+  }
+
+  @Override
+  public SerDeStats getSerDeStats() {
+    return stats;
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java:153]: This `stats` is never updated. You need to either maintain a local variable to track the metrics and update the real into `stats` here, or update `stats` every time `deserialize` called.

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  public Object deserialize(Writable writable) throws SerDeException
+  {
+    ThriftGenericRow row = (ThriftGenericRow) ((ThriftWritable) writable).get();
+    try {
+      row.parse();
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java:161]: You can actually delay the "real" deserialization until either `getStructFieldData` or `getStructFieldsDataAsList` is called.

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGenericRow.java [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    len = trans.getBufferPosition() - off;
+  }
+
+  public void parse()
+      throws TException
+  {
+    parse(null);
+  }
+
+  public void parse(short[] thriftIds)
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGenericRow.java]: You can actually change this method to incrementally parsing the fields that required by `ThriftGeneralObjectInspector.getStructFieldData`.

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+package org.apache.hadoop.hive.ql.io.thrift.twitter;
+
+import java.util.List;
+
+import org.apache.hadoop.hive.serde2.BaseStructObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.StructField;
+
+public class ThriftGeneralObjectInspector extends BaseStructObjectInspector {
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java]: Rename to `ThriftStructObjectInspector`

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGenericRow.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+        return null;
+    }
+  }
+
+  private Object readStruct(TProtocol iprot)
+      throws TException
+  {
+    ThriftGenericRow elem = new ThriftGenericRow();
+    elem.read(iprot);
+    elem.parse();
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGenericRow.java:155]: You may don't have to run `elem.parse()` here because Hive will hierarchically walk down to every nested level so that you would only need to process one level per `getStructFieldData` or `getStructFieldsDataAsList` call.

**Review by Yaliang — CHANGES_REQUESTED** [review]
Good work. I left some comments.

Beside the comments, I believe you haven't been looking into the optimization like selective deserialization. If you are interested on improving the performance, you might need to checkout `ColumnProjectionUtils.READ_NESTED_COLUMN_PATH_CONF_STR` configuration in order to prune the TypeInfo when you initialize `ThriftGeneralDeserializer`.

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java (hunk 2 above)
Comment (mghosh4) [reply ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java]: Fixed

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java (hunk 3 above)
Comment (mghosh4) [reply ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralObjectInspector.java]: Fixed

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralInputFormat.java (hunk 4 above)
Comment (mghosh4) [reply ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralInputFormat.java]: We do not need to. I was just wondering if later people start using a different TBase Class in Table definition. At that point using ThriftGenericRow might not be correct. It might still work and fail somewhere else. Do you think that is a valid concern?

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Table.java (hunk 6 above)
Comment (mghosh4) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/Table.java]: Yes I was reading about storage handler recently. Right now we do not use Storage Handlers with HIVE. Most likely, we will not come. Do you want to keep the check as a safeguard or let it fail if ever and fix it at that point

File: metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreUtils.java (hunk 7 above)
Comment (mghosh4) [reply metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreUtils.java:436]: It is also called by Table::getDeserializerFromMetaStore method

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java (hunk 1 above)
Comment (mghosh4) [reply ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGeneralDeserializer.java:135]: HIVE support C style union where it can contain data of any type but only one field at a time. I am not sure thrift supports something like that. So leaving it empty makes sense.

**Review by mghosh4 — COMMENTED** [review]
Fixed most of the major comments

File: ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGenericRow.java [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+        TMemoryInputTransport trans = new TMemoryInputTransport(buf, fieldOffsets.get(thriftId), len);
+        TBinaryProtocol iprot = new TBinaryProtocol(trans);
+
+        TField field = iprot.readFieldBegin();
+        fieldValues.put(field.id, readElem(iprot, field.type));
+        fieldDeserialized.add(field.id);
+      }
+      catch (TException e)
+      {
+        log.error("Thrift Id missing in ThriftGenericRow");
```
Comment (Yaliang) [inline ql/src/java/org/apache/hadoop/hive/ql/io/thrift/twitter/ThriftGenericRow.java]: you can use a general error message instead. And include the `e` in logger can help to further debug

```
log.error("Failed to decode the field with thrift id: " + thriftId + ", the field may be returned as null", e);
```

File: ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+          .setTweetType(TweetType.REPLY)
+          .setFullId(1234567)
+          .setPic("abc".getBytes())
+          .setAttr(ImmutableMap.of("a", "a"));
+      TMemoryBuffer transport = new TMemoryBuffer(4096);
+      TBinaryProtocol protocol = new TBinaryProtocol(transport);
+      tweet.write(protocol);
+
+      // Create the SerDe
+      System.out.println("test: testThriftDeserializer");
```
Comment (Yaliang) [inline ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java]: remove or use a logger :)

File: ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+      Tweet tweet = new Tweet(1, "newUser", "hello world")
+          .setLoc(new Location(1234, 5678))
+          .setAge((short) 26)
+          .setB((byte) 10)
+          .setIsDeleted(false)
+          .setTweetType(TweetType.REPLY)
+          .setFullId(1234567)
+          .setPic("abc".getBytes())
+          .setAttr(ImmutableMap.of("a", "a"));
+      TMemoryBuffer transport = new TMemoryBuffer(4096);
```
Comment (Yaliang) [inline ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java]: no longer use line 69-71

File: ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+      }
+    }
+
+    return true;
+  }
+
+  private Properties createProperties() {
+    StructTypeInfo tweepStruct = (StructTypeInfo) TypeInfoFactory.getStructTypeInfo(
+        toList("userid", "username", "text", "loc", "tweettype", "isdeleted", "b", "age", "fullid", "pic", "attr", "items", "language"),
+        toList(TypeInfoFactory.getPrimitiveTypeInfo(serdeConstants.INT_TYPE_NAME),
```
Comment (Yaliang) [inline ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java]: this block is a little bit wordy, do you want to have static imports of those type names to save `serdeConstants.` and also have a tiny helper function to simplify 
`TypeInfoFactory.getPrimitiveTypeInfo(xxx)`?

File: ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+    // Deserialize
+    final Object row = serDe.deserialize(writable);
+    assertEquals("deserialization gives the wrong object class", row.getClass(), ThriftGenericRow.class);
+    assertEquals("size incorrect after deserialization", serDe.getSerDeStats().getRawDataSize(), ((ThriftGenericRow) writable.get()).length());
+
+    // Lazy Field Deserialize
+    ThriftStructObjectInspector oi = (ThriftStructObjectInspector)serDe.getObjectInspector();
+    assertEquals("This test should pass", true, deserializeAndTestAllFields(row, oi, tweet, true));
+
+    Tweet tweet1 = new Tweet(1, "newUser", "hello world")
```
Comment (Yaliang) [inline ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java]: I think you need to have another method that tests this `tweet1`, as you have done for `tweet`.

File: ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+        .setTweetType(TweetType.REPLY)
+        .setFullId(1234568)
+        .setPic("abc".getBytes())
+        .setAttr(ImmutableMap.of("a", "a"));
+
+    assertEquals("This test should not pass", false, deserializeAndTestAllFields(row, oi, tweet1, false));
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T> boolean deserializeAndTestAllFields(Object row, ObjectInspector oi, T expectedResult, boolean testLazyDeserialize)
```
Comment (Yaliang) [inline ql/src/test/org/apache/hadoop/hive/ql/io/thrift/twitter/TestThriftGeneralDeserializer.java]: Maybe rename `deserializeAndTestAllFields`  to `assertResult` and rename the first parameter as `Object result`? About the return value, I think it's actually fine to use assert methods within this method. In that case, if you want to test inequal value like `tweet1`, you can catch an assertion error and try to test whether the assertion error message is the same as what you expected.

**Review by Yaliang — COMMENTED** [review]
Nice work! I added some comments.

**Review by Yaliang — APPROVED** [review]

**Review by Yaliang — APPROVED** [review]
